internal/assets/tokens_transferred.go

```diff
@@ -0,0 +1,66 @@
+// Copyright © 2021 Kaleido, Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package assets
+
+import (
+	"context"
+
+	"github.com/hyperledger/firefly/internal/log"
+	"github.com/hyperledger/firefly/pkg/fftypes"
+	"github.com/hyperledger/firefly/pkg/tokens"
+)
+
+func (am *assetManager) TokensTransferred(tk tokens.Plugin, transfer *fftypes.TokenTransfer, signingIdentity string, protocolTxID string, additionalInfo fftypes.JSONObject) error {
+	return am.retry.Do(am.ctx, "persist token pool", func(attempt int) (bool, error) {
+		err := am.database.RunAsGroup(am.ctx, func(ctx context.Context) error {
+			pool, err := am.database.GetTokenPoolByProtocolID(ctx, transfer.PoolProtocolID)
+			if err != nil {
+				return err
+			}
+			if pool == nil {
+				log.L(ctx).Warnf("Token transfer received for unknown pool '%s' - ignoring: %s", transfer.PoolProtocolID, protocolTxID)
+				return nil
+			}
+			if err := am.database.UpsertTokenTransfer(ctx, transfer); err != nil {
+				log.L(ctx).Errorf("Failed to record token transfer '%s': %s", transfer.ProtocolID, err)
+				return err
+			}
```

Comment:
The logic below is relying strongly on the transactional nature of the database in order to end up with the correct account balances. Just want to confirm that's OK.

The ERC1155 plugin actually allows balance querying, so we could include the ending balance with each transfer event... but that seems more complex on the plugin side and probably won't be possible with other plugins, e.g. a UTXO model.